### PR TITLE
ENH: Palette image write support for TIFF and PNG.

### PR DESCRIPTION
**Pull request by YannLePoul**
For images with one component, user can now write the image as 1 component pixels and a color palette.

## PR Checklist

- [X] :no_entry_sign: [Makes breaking changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes)
- [X] :no_entry_sign: [Makes design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes)
- [X] :no_entry_sign: Adds the License notice to new files.
- [] :no_entry_sign: Adds Python wrapping.
- [X] :no_entry_sign: Adds tests and baseline comparison (quantitative).
- [X] :no_entry_sign: [Adds test data](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/UploadBinaryData.md).
- [] :no_entry_sign: Adds Examples to [ITKExamples](https://github.com/InsightSoftwareConsortium/ITKExamples)
- [] :no_entry_sign: Adds Documentation.

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by YannLePoul** [comment]
I added some testing  for palette writing, and improve the clarity and make test more comprehensive.

**Comment by YannLePoul** [comment]
only ITKIOPNGKWStyleTest is failing, error is Header mismatch for itkPNGImageIO.cxx, but I didn't touch the header.

File: Modules/IO/PNG/include/itkPNGImageIO.h [hunk 1]
```diff
@@ -59,6 +59,9 @@ class ITKIOPNG_EXPORT PNGImageIO : public ImageIOBase
    * 0 is returned */
   itkGetConstReferenceMacro(ColorPalette, PaletteType);
 
+  /** Set the palette of the image. */
+  itkSetMacro(ColorPalette, PaletteType);
```
Comment (dzenanz) [inline Modules/IO/PNG/include/itkPNGImageIO.h]: [SetMacro](https://github.com/InsightSoftwareConsortium/ITK/blob/2074c2f9e5ed3f087d0e2059f8e0e8992fcad7ef/Modules/Core/Common/include/itkMacro.h#L873) calls [Debug macro](https://github.com/InsightSoftwareConsortium/ITK/blob/2074c2f9e5ed3f087d0e2059f8e0e8992fcad7ef/Modules/Core/Common/include/itkMacro.h#L396), which attempts to `ostream<<x`. This is not defined for `std::vector`.

File: Modules/IO/TIFF/src/itkTIFFImageIO.cxx [hunk 2]
```diff
@@ -974,6 +1000,52 @@ TIFFImageIO::PopulateColorPalette()
...
 
+void
+TIFFImageIO::AllocateTiffPalette(int bps)
+{
+  m_ColorRed = nullptr;
+  m_ColorGreen = nullptr;
+  m_ColorBlue = nullptr;
+
+  uint16 bpp = static_cast<uint16>(bps);
+  m_ColorRed = static_cast<uint16 *>(_TIFFmalloc(1 << bpp * sizeof(uint16)));
```
Comment (dzenanz) [inline Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: `D:\a\1\s\Modules\IO\TIFF\src\itkTIFFImageIO.cxx(1011): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)` and a [few more times](https://open.cdash.org/viewBuildError.php?type=1&buildid=6087738).

File: Modules/IO/PNG/include/itkPNGImageIO.h (hunk 1 above)
Comment (YannLePoul) [reply Modules/IO/PNG/include/itkPNGImageIO.h]: Sorry, I didn't got any compilation error on my machine. How should I fix this ? Not using the macro ?

Comment (dzenanz) [reply Modules/IO/PNG/include/itkPNGImageIO.h]: Yes, make a code similar to the macro. Maybe copy-paste and remove the debug output statement?

File: Modules/IO/TIFF/src/itkTIFFImageIO.cxx (hunk 2 above)
Comment (YannLePoul) [reply Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: I think I've stolen  this code from [here](https://www.asmail.be/msg0055228619.html). [Another](http://maptools-org.996276.n3.nabble.com/Adding-alpha-to-palette-files-td13203.html) example.  I think here, it is mainly used for computing fast power of 2 `1 << (bpp * sizeof(uint16)) == std::pow(2, (bpp * sizeof(uint16))) `.  So 64 or 32 bit shift would give the same result. I could just static cast the result to a size t (input type of TIFFmalloc).

Comment (dzenanz) [reply Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: > computing fast power of 2

That is clear. But let's get rid of the warning in the least intrusive way. Maybe make `bpp` variable `uint8`?

Comment (YannLePoul) [reply Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: Would uint8 it make a difference ? What about :
```
  tmsize_t array_size = 1 << bpp * sizeof(uint16);
  m_ColorRed = static_cast<uint16 *>(_TIFFmalloc(array_size));
  if (!m_ColorRed)
  {
    _TIFFfree(m_ColorRed);
    itkExceptionMacro("Can't allocate space for Red channel of component tables.");
  }
```
It is anyway computed 3 times. I don't see any warning when I compile on my machine. Do you have any advice to see those ?

Comment (dzenanz) [reply Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: When configuring ITK, CMAKE_CXX_FLAGS is set to `/DWIN32 /D_WINDOWS /W3 /GR /EHsc` by default. Change `/W3` into `/W4` to increase warning level.

**Comment by dzenanz** [comment]
Windows test failure seems unrelated.

**Comment by YannLePoul** [comment]
Is there anything left to do ?

File: Modules/IO/TIFF/src/itkTIFFImageIO.cxx [hunk 3]
```diff
@@ -974,6 +1000,54 @@ TIFFImageIO::PopulateColorPalette()
...
+  m_ColorRed = static_cast<uint16 *>(_TIFFmalloc(array_size));
+  if (!m_ColorRed)
+  {
+    _TIFFfree(m_ColorRed);
+    itkExceptionMacro("Can't allocate space for Red channel of component tables.");
+  }
+  m_ColorGreen = static_cast<uint16 *>(_TIFFmalloc(array_size));
+  if (!m_ColorGreen)
+  {
+    _TIFFfree(m_ColorGreen);
```
Comment (dzenanz) [inline Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: Shouldn't we free red channel here too?

File: Modules/IO/TIFF/src/itkTIFFImageIO.cxx [hunk 4]
```diff
@@ -974,6 +1000,54 @@ TIFFImageIO::PopulateColorPalette()
...
+    itkExceptionMacro("Can't allocate space for Green channel of component tables.");
+  }
+  m_ColorBlue = static_cast<uint16 *>(_TIFFmalloc(array_size));
+  if (!m_ColorBlue)
+  {
+    _TIFFfree(m_ColorBlue);
```
Comment (dzenanz) [inline Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: Free also red and green?

File: Modules/IO/TIFF/src/itkTIFFImageIO.cxx (hunk 3 above)
Comment (YannLePoul) [reply Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: yes, good catch!

File: Modules/IO/TIFF/src/itkTIFFImageIO.cxx (hunk 4 above)
Comment (YannLePoul) [reply Modules/IO/TIFF/src/itkTIFFImageIO.cxx]: yep

**Pull request merged**